### PR TITLE
lightning: add new configuration for incremental restore

### PR DESCRIPTION
**Pull request by glorv**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: tidb-lightning/tidb-lightning-configuration.md [hunk 1]
```diff
@@ -95,6 +99,8 @@ driver = "file"
 [tikv-importer]
 # 选择后端：“local” 或 “importer” 或 “tidb”
 # backend = "local"
+# 是否开启并行导入模式
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-configuration.md:102]:
```suggestion
# 是否开启并行导入模式。“true” 为开启并行导入模式，"false" 为关闭并行导入模式。默认为 "false"。
```

File: tidb-lightning/tidb-lightning-configuration.md [hunk 2]
```diff
@@ -59,6 +59,10 @@ table-concurrency = 6
 # 对于不同的存储介质，此参数可能需要调整以达到最佳效率。
 io-concurrency = 5
 
+# 在并行导入模式下，在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名字，默认为 "lightning_metadata"
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-configuration.md:62]:
```suggestion
# 在并行导入模式下，在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名字。默认为 "lightning_metadata"。
```

File: tidb-lightning/tidb-lightning-configuration.md [hunk 3]
```diff
@@ -59,6 +59,10 @@ table-concurrency = 6
...
+# 如果未开启并行导入模式，无须设置此配置项。
+meta-schema-name = "lightning_metadata"
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-configuration.md:64]: 如果有多个 Lightning 实例，是否需要添加多个 meta-schema-name ？可以给个多个实例的例子，毕竟启动并行导入的时候肯定有多个实例。

File: tidb-lightning/tidb-lightning-configuration.md (hunk 1 above)
Comment (glorv) [reply tidb-lightning/tidb-lightning-configuration.md:102]: 👍。true/false 为 bool 类型，不需要加 "" 😅。

File: tidb-lightning/tidb-lightning-prechecks.md [hunk 4]
```diff
@@ -9,12 +9,13 @@ summary: 本文档介绍了 TiDB Lightning 前置检查功能，确保 TiDB Ligh
...
+|  检查项 | 支持版本 | 解释 |
+|  ----  |  --- | ----  |
+| 集群版本/状态是否正常| >= 5.3.0 | 检查配置中集群是否可以连接，当后端模式是 Local 的时候，还会检查 TiKV/PD/TiFlash 版本是否支持 Local 导入模式。|
+| 是否有权限读取数据 | >= 5.3.0 | 检查当从云存储（Amazon S3）读取数据的时候，是否有对应的权限，确保不会因权限缺失导致导入中断。|
+| 导入空间是否足够 | >= 5.3.0 | 检查 TiKV 集群是否有足够空间导入数据。检查时会对数据源进行采样，通过采样结果预估索引大小占比。由于估算中考虑了索引，因此可能会出现尽管数据源大小低于本地盘可用空间，但依然无法通过检测的情况。当后端是 Local 的时候，因为需要在本地进行外部排序，所以还会检查本地存储是否足够。| 
+| region 分布状态 | >= 5.3.0 | 检查 TiKV 集群的 region 分布是否均匀，以及是否存在大量空 region，如果空 region 的数量大于 max(1000,  表的数量 * 3) ，即大于 “1000” 和 “3 倍表数量”二者中的最大者，无法执行导入。 |
+| 数据文件是否有大 CSV 文件 | >= 5.3.0 | 当备份文件中出现大于 10 GiB 的 CSV 文件且无法进行自动切分(StrictFormat=false)的时候，会导致导入性能下降。该检查的目的是提醒用户确保数据格式的情况下，开启自动切分 CSV 功能。 |
+| 是否可以从断点恢复 | >= 5.3.0 | 该检查是确保断点恢复过程中，不会出现对源文件和数据库中 schema 进行修改，导致导入错误数据的情况。|
+| 是否可以导入数据到已存在的数据表中 | >= 5.3.0 | 当导入到已创建好的数据表中时，该检查尽可能的检查此次导入的源文件是否和已存在的数据表匹配。检查列数是否匹配，如果源文件存在列名，则检查列名是否匹配。当源文件存在缺省列，则检查缺省列在数据表中是否存在 Default Value，如果存在，则检查通过。|
+| 导入的目标表是否为空 | >= 5.4.0 | 默认模式下，如果导入的目标表不为空，则 TiDB Lightning 会自动报错退出；如果开启并行导入模式，则会跳过此检查项。 |
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-prechecks.md:21]:
```suggestion
| 导入的目标表是否为空 | >= 5.4.0 | 如果未开启并行导入模式，如果导入的目标表中存在数据，则 TiDB Lightning 会自动报错退出；如果开启并行导入模式，则会跳过此检查项。 |
```

File: tidb-lightning/tidb-lightning-configuration.md [hunk 5]
```diff
@@ -95,6 +103,8 @@ driver = "file"
 [tikv-importer]
 # 选择后端：“local” 或 “importer” 或 “tidb”
 # backend = "local"
+# 是否开启并行导入模式。true 表示开启并行导入模式，false 表示关闭并行导入模式。默认为 false。
```
Comment (sunzhaoyang) [inline tidb-lightning/tidb-lightning-configuration.md:106]: 我感觉这里可能会让用户疑惑，参数名是 incremental-import ，注释里讲的是并行导入。其实刚开始斟酌这个参数名的时候也是纠结是否直接用并行导入含义的开关，但最终还是决定用 incremental-import 更加符合这个行为本身。建议注释再详细说明一下：

```
# 是否允许向已存在数据的表内导入数据。
# Lightning 一般用于初始导入数据，默认值 false 表示目标表存在数据时禁止导入，true 则表示可以导入
# 当使用并行导入模式时，多个 Lightning 进程同时导入则要求此开关必须为 true
```

**Comment by sunzhaoyang** [comment]
rest LGTM

File: tidb-lightning/tidb-lightning-configuration.md (hunk 5 above)
Comment (glorv) [reply tidb-lightning/tidb-lightning-configuration.md:106]: 那感觉还需要增加一个对增量导入这种模式的单独说明😂。否则用户还是会比较疑惑

Comment (sunzhaoyang) [inline tidb-lightning/tidb-lightning-configuration.md:106]:
```suggestion
# 是否允许向已存在数据的表导入数据。
# Lightning 一般用于初始导入数据，因此目标表要求为空。若目标表存在数据，默认值 false 表示禁止导入，true 则表示关闭此限制继续导入。
# 当使用并行导入模式时，由于多个 Lightning 进程同时导入一张表，因此要求此开关必须为 true
```

**Comment by gozssky** [comment]
https://github.com/pingcap/docs-cn/blob/master/tidb-lightning/tidb-lightning-distributed-import.md This document also needs to update.

File: tidb-lightning/tidb-lightning-configuration.md [hunk 6]
```diff
@@ -59,6 +59,14 @@ table-concurrency = 6
 # 对于不同的存储介质，此参数可能需要调整以达到最佳效率。
 io-concurrency = 5
 
+# 在并行导入模式下，在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名字，默认为 "lightning_metadata"。
+# 如果未开启并行导入模式，无须设置此配置项。
+# 注意：
+# 1. 对于同一批的并行导入的每个 TiDB Lightning 实例，必须将此配置项设置为相同的值，否则将无法确保导入的正确性。
+# 2. 如果开启并行导入模式，需要确保导入使用的用户(对于 tidb.user 配置项)有权限创建和访问此配置对于的库。
+# 3. TiDB Lightning 在导入完成后会删除此 schema，因此最好不要将此配置设置为一个已存在的库名。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-configuration.md:67]:
```suggestion
# **注意：**
# 1. 对于参与同一批并行导入的每个 TiDB Lightning 实例，必须将此配置项设置为相同的值，否则将无法确保导入数据的正确性。
# 2. 如果开启并行导入模式，需要确保导入使用的用户（对于 tidb.user 配置项）有权限创建和访问此配置对应的库。
# 3. TiDB Lightning 在导入完成后会删除此 schema，因此不要使用已存在的库名配置该参数。
```

File: tidb-lightning/tidb-lightning-configuration.md [hunk 7]
```diff
@@ -95,6 +103,9 @@ driver = "file"
 [tikv-importer]
 # 选择后端：“local” 或 “importer” 或 “tidb”
 # backend = "local"
+# 是否允许向已存在数据的表导入数据。默认值为 false。
+# 当使用并行导入模式时，由于多个 Lightning 实例同时导入一张表，因此要求此开关必须为 true。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-configuration.md:107]:
```suggestion
# 当使用并行导入模式时，由于多个 TiDB Lightning 实例同时导入一张表，因此此开关必须设置为 true。 
```

File: tidb-lightning/tidb-lightning-configuration.md [hunk 8]
```diff
@@ -59,6 +59,14 @@ table-concurrency = 6
 # 对于不同的存储介质，此参数可能需要调整以达到最佳效率。
 io-concurrency = 5
 
+# 在并行导入模式下，在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名字，默认为 "lightning_metadata"。
+# 如果未开启并行导入模式，无须设置此配置项。
+# **注意：**
+# 1. 对于参与同一批并行导入的每个 TiDB Lightning 实例，必须将此配置项设置为相同的值，否则将无法确保导入数据的正确性。
+# 2. 如果开启并行导入模式，需要确保导入使用的用户（对于 tidb.user 配置项）有权限创建和访问此配置对应的库。
+# 3. TiDB Lightning 在导入完成后会删除此 schema，因此不要使用已存在的库名配置该参数。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-configuration.md:67]:
```suggestion
# - 对于参与同一批并行导入的每个 TiDB Lightning 实例，必须将此配置项设置为相同的值，否则将无法确保导入数据的正确性。
# - 如果开启并行导入模式，需要确保导入使用的用户（对于 tidb.user 配置项）有权限创建和访问此配置对应的库。
# - TiDB Lightning 在导入完成后会删除此 schema，因此不要使用已存在的库名配置该参数。
```

File: tidb-lightning/tidb-lightning-configuration.md [hunk 9]
```diff
@@ -59,6 +59,14 @@ table-concurrency = 6
 # 对于不同的存储介质，此参数可能需要调整以达到最佳效率。
 io-concurrency = 5
 
+# 在并行导入模式下，在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名字，默认为 "lightning_metadata"。
+# 如果未开启并行导入模式，无须设置此配置项。
+# **注意：**
+# - 对于参与同一批并行导入的每个 TiDB Lightning 实例，必须将此配置项设置为相同的值，否则将无法确保导入数据的正确性。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-configuration.md:65]:
```suggestion
# - 对于参与同一批并行导入的每个 TiDB Lightning 实例，该参数设置的值必须相同，否则将无法确保导入数据的正确性。
```

**Comment by glorv** [comment]
> https://github.com/pingcap/docs-cn/blob/master/tidb-lightning/tidb-lightning-distributed-import.md This document also needs to update.

done.

File: tidb-lightning/tidb-lightning-distributed-import.md [hunk 10]
```diff
@@ -80,11 +80,14 @@ TiDB Lightning 并行导入可以用于以下场景：
 [lightning]
 status-addr = ":8289"
 
+
 [mydumper]
 # 设置为 Dumpling 导出数据的路径，如果 Dumpling 执行了多次并分属不同的目录，请将多次导出的数据置放在相同的父目录下并指定此父目录即可。
 data-source-dir = "/path/to/source-dir"
 
 [tikv-importer]
+# 开启并行导入功能
```
Comment (sunzhaoyang) [inline tidb-lightning/tidb-lightning-distributed-import.md:89]:
```suggestion
# 是否允许向已存在数据的表导入数据。默认值为 false。
 # 当使用并行导入模式时，由于多个 TiDB Lightning 实例同时导入一张表，因此此开关必须设置为 true。 
```

File: tidb-lightning/tidb-lightning-distributed-import.md [hunk 11]
```diff
@@ -80,11 +80,14 @@ TiDB Lightning 并行导入可以用于以下场景：
...
+incremental-import = "true"
 # 使用 Local 后端
```
Comment (sunzhaoyang) [inline tidb-lightning/tidb-lightning-distributed-import.md:91]:
```suggestion
# "local"：默认使用该模式，适用于 TB 级以上大数据量，但导入期间下游 TiDB 无法对外提供服务。
# "tidb"：TB 级以下数据量也可以采用`tidb`后端模式，下游 TiDB 可正常提供服务。 关于后端模式更多信息请参阅：https://docs.pingcap.com/tidb/stable/tidb-lightning-backends
```

File: tidb-lightning/tidb-lightning-configuration.md [hunk 12]
```diff
@@ -95,6 +103,9 @@ driver = "file"
 [tikv-importer]
 # 选择后端：“local” 或 “importer” 或 “tidb”
```
Comment (sunzhaoyang) [inline tidb-lightning/tidb-lightning-configuration.md:104]:
```suggestion
# "local"：默认使用该模式，适用于 TB 级以上大数据量，但导入期间下游 TiDB 无法对外提供服务。
# "tidb"：TB 级以下数据量也可以采用`tidb`后端模式，下游 TiDB 可正常提供服务。 关于后端模式更多信息请参阅：https://docs.pingcap.com/tidb/stable/tidb-lightning-backends```

**Comment by sunzhaoyang** [comment]
rest LGTM

File: tidb-lightning/tidb-lightning-distributed-import.md (hunk 11 above)
Comment (TomShawn) [reply tidb-lightning/tidb-lightning-distributed-import.md:91]: 在配置文件注释里建议不放链接，因为用户不能直接点开，而且存在 URL 版本问题

File: tidb-lightning/tidb-lightning-prechecks.md [hunk 13]
```diff
@@ -9,12 +9,13 @@ summary: 本文档介绍了 TiDB Lightning 前置检查功能，确保 TiDB Ligh
...
-| 数据文件是否有大 CSV 文件 | 当备份文件中出现大于 10 GiB 的 CSV 文件且无法进行自动切分(StrictFormat=false)的时候，会导致导入性能下降。该检查的目的是提醒用户确保数据格式的情况下，开启自动切分 CSV 功能。 |
-| 是否可以从断点恢复 | 该检查是确保断点恢复过程中，不会出现对源文件和数据库中 schema 进行修改，导致导入错误数据的情况。|
-| 是否可以导入数据到已存在的数据表中| 当导入到已创建好的数据表中时，该检查尽可能的检查此次导入的源文件是否和已存在的数据表匹配。检查列数是否匹配，如果源文件存在列名，则检查列名是否匹配。当源文件存在缺省列，则检查缺省列在数据表中是否存在 Default Value，如果存在，则检查通过。|
\ No newline at end of file
+|  检查项 | 支持版本 | 解释 |
+|  ----  |  --- | ----  |
+| 集群版本/状态是否正常| >= 5.3.0 | 检查配置中集群是否可以连接，当后端模式是 Local 的时候，还会检查 TiKV/PD/TiFlash 版本是否支持 Local 导入模式。|
+| 是否有权限读取数据 | >= 5.3.0 | 检查当从云存储（Amazon S3）读取数据的时候，是否有对应的权限，确保不会因权限缺失导致导入中断。|
+| 导入空间是否足够 | >= 5.3.0 | 检查 TiKV 集群是否有足够空间导入数据。检查时会对数据源进行采样，通过采样结果预估索引大小占比。由于估算中考虑了索引，因此可能会出现尽管数据源大小低于本地盘可用空间，但依然无法通过检测的情况。当后端是 Local 的时候，因为需要在本地进行外部排序，所以还会检查本地存储是否足够。| 
+| region 分布状态 | >= 5.3.0 | 检查 TiKV 集群的 region 分布是否均匀，以及是否存在大量空 region，如果空 region 的数量大于 max(1000,  表的数量 * 3) ，即大于 “1000” 和 “3 倍表数量”二者中的最大者，无法执行导入。 |
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-prechecks.md:17]:
```suggestion
| Region 分布状态 | >= 5.3.0 | 检查 TiKV 集群的 Region 分布是否均匀，以及是否存在大量空 region，如果空 Region 的数量大于 `max(1000,  表的数量 * 3)`，即大于 "1000" 和 "3 倍表数量"二者中的较大者，TiDB Lightning 无法执行导入。 |
```

File: tidb-lightning/tidb-lightning-prechecks.md [hunk 14]
```diff
@@ -9,12 +9,13 @@ summary: 本文档介绍了 TiDB Lightning 前置检查功能，确保 TiDB Ligh
...
+| 数据文件是否有大 CSV 文件 | >= 5.3.0 | 当备份文件中出现大于 10 GiB 的 CSV 文件且无法进行自动切分(StrictFormat=false)的时候，会导致导入性能下降。该检查的目的是提醒用户确保数据格式的情况下，开启自动切分 CSV 功能。 |
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-prechecks.md:18]:
```suggestion
| 数据文件是否有大 CSV 文件 | >= 5.3.0 | 当备份文件中出现大于 10 GiB 的 CSV 文件且无法进行自动切分 (StrictFormat=false) 的时候，会导致导入性能下降。该检查的目的是提醒用户确保数据格式的情况下，开启自动切分 CSV 功能。 |
```

File: tidb-lightning/tidb-lightning-configuration.md (hunk 12 above)
Comment (TomShawn) [reply tidb-lightning/tidb-lightning-configuration.md:104]: 在配置文件注释里建议不放链接，因为用户不能直接点开，而且存在 URL 版本问题

File: tidb-lightning/tidb-lightning-distributed-import.md [hunk 15]
```diff
@@ -80,12 +80,17 @@ TiDB Lightning 并行导入可以用于以下场景：
...
 
+
 [mydumper]
 # 设置为 Dumpling 导出数据的路径，如果 Dumpling 执行了多次并分属不同的目录，请将多次导出的数据置放在相同的父目录下并指定此父目录即可。
 data-source-dir = "/path/to/source-dir"
 
 [tikv-importer]
-# 使用 Local 后端
+# 是否允许向已存在数据的表导入数据。默认值为 false。
+ # 当使用并行导入模式时，由于多个 TiDB Lightning 实例同时导入一张表，因此此开关必须设置为 true。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-distributed-import.md:90]:
```suggestion
# 当使用并行导入模式时，由于多个 TiDB Lightning 实例同时导入一张表，因此此开关必须设置为 true。 
```

File: tidb-lightning/tidb-lightning-distributed-import.md [hunk 16]
```diff
@@ -80,12 +80,17 @@ TiDB Lightning 并行导入可以用于以下场景：
...
+
 [mydumper]
 # 设置为 Dumpling 导出数据的路径，如果 Dumpling 执行了多次并分属不同的目录，请将多次导出的数据置放在相同的父目录下并指定此父目录即可。
 data-source-dir = "/path/to/source-dir"
 
 [tikv-importer]
-# 使用 Local 后端
+# 是否允许向已存在数据的表导入数据。默认值为 false。
+# 当使用并行导入模式时，由于多个 TiDB Lightning 实例同时导入一张表，因此此开关必须设置为 true。 
+incremental-import = "true"
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-distributed-import.md:91]:
```suggestion
incremental-import = true
```

File: tidb-lightning/tidb-lightning-distributed-import.md [hunk 17]
```diff
@@ -80,12 +80,17 @@ TiDB Lightning 并行导入可以用于以下场景：
 [lightning]
 status-addr = ":8289"
 
+
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-distributed-import.md:83]:
```suggestion
```

**Review by TomShawn — APPROVED** [review]
LGTM

**Comment by TomShawn** [comment]
@sunzhaoyang @gozssky PTAL again, thanks!

**Review by hfxsd — APPROVED** [review]
LGTM

**Pull request merged**